tendermint/src/abci/types.rs

```diff
@@ -0,0 +1,306 @@
+pub struct Validator {
+    /// The validator's address (the first 20 bytes of `SHA256(public_key)`).
+    pub address: [u8; 20],
+    /// The voting power of the validator.
+    pub power: i64,
+}
+
+/// A change to the validator set.
+///
+/// Used to inform Tendermint of changes to the validator set.
+///
+/// [ABCI documentation](https://docs.tendermint.com/master/spec/abci/abci.html#validatorupdate)
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct ValidatorUpdate {
+    /// The validator's public key.
+    pub pub_key: PublicKey,
+    /// The validator's voting power.
+    pub power: i64,
+}
+
+/// Information about a whether a validator signed the last block.
+///
+/// [ABCI documentation](https://docs.tendermint.com/master/spec/abci/abci.html#voteinfo)
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct VoteInfo {
+    /// Identifies the validator.
+    pub validator: Validator,
+    /// Whether or not the validator signed the last block.
+    pub signed_last_block: bool,
+}
+
+/// The possible kinds of [`Evidence`].
+///
+/// Note: the
+/// [ABCI documentation](https://docs.tendermint.com/master/spec/abci/abci.html#evidencetype-2)
+/// calls this `EvidenceType`, but we follow the Rust convention and name it `EvidenceKind`
+/// to avoid confusion with Rust types.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+#[repr(i32)]
+pub enum EvidenceKind {
+    /// Unknown evidence type (proto default value).
+    Unknown = 0,
+    /// Evidence that the validator voted for two different blocks in the same
+    /// round of the same height.
+    DuplicateVote = 1,
+    /// Evidence that a validator attacked a light client.
+    LightClientAttack = 2,
+}
+
+/// Evidence of validator misbehavior.
+///
+/// [ABCI documentation](https://docs.tendermint.com/master/spec/abci/abci.html#evidence)
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct Evidence {
+    /// The kind of evidence.
+    ///
+    /// Note: this field is called `type` in the protobuf, but we call it `kind`
+    /// to avoid the Rust keyword.
+    pub kind: EvidenceKind,
+    /// The offending validator.
+    pub validator: Validator,
+    /// The height when the offense occurred.
+    pub height: i64,
+    /// The corresponding time when the offense occurred.
+    pub time: DateTime<Utc>,
+    /// Total voting power of the validator set at `height`.
+    ///
+    /// This is included in case the ABCI application does not store historical
+    /// validators, cf.
+    /// [#4581](https://github.com/tendermint/tendermint/issues/4581)
+    pub total_voting_power: i64,
+}
+
+/// Information on the last block commit.
+///
+/// [ABCI documentation](https://docs.tendermint.com/master/spec/abci/abci.html#lastcommitinfo)
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct LastCommitInfo {
+    /// The commit round.
+    ///
+    /// Reflects the total number of rounds it took to come to consensus for the
+    /// current block.
+    pub round: i32,
```

Comment:
See [`Round`](https://github.com/informalsystems/tendermint-rs/blob/5bf6bae177cebb32bd9926f300d6e42edc23041c/tendermint/src/block/round.rs#L13).

Reply:
Fixed in fde60809